geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/metadata/FileBasedMetadata.scala

```diff
@@ -328,37 +340,65 @@ class FileBasedMetadata(
     }
   }
 
+  /**
+   * Delete a seq of paths
+   *
+   * @param paths paths to delete
+   * @param threads number of threads to use
+   */
+  private def delete(paths: Iterable[Path], threads: Int): Unit = {
+    if (threads < 2) {
+      paths.foreach(fc.delete(_, false))
+    } else {
+      val ec = ExecutionContext.fromExecutorService(new CachedThreadPool(threads))
+      try {
+        val parPaths = paths.par
+        parPaths.tasksupport = new ExecutionContextTaskSupport(ec)
+        parPaths.foreach(fc.delete(_, false))
+      } finally {
+        ec.shutdown()
+      }
+    }
+  }
+
   private class DirectoryWorker(
       es: ExecutorService,
       phaser: Phaser,
-      dir: Path,
+      listDirectory: => RemoteIterator[FileStatus],
       result: ConcurrentHashMap[String, PartitionFiles]
     ) extends Runnable {
 
     override def run(): Unit = {
       try {
-        val iter = fc.listStatus(dir)
-        while (iter.hasNext) {
+        var i = phaser.getRegisteredParties + 1
```

Comment:
i think registered parties will always be 1 in this case. there are various checks to ensure that it does not end up being MaxParties, if it did then we wouldn't be able to create a new child phaser and things would break...

Reply:
What all would be needed for that to happen?  Would we have to have 65K * 65K files or something like that?

Separate question, do we need to help people with 4 billion+ files in a directory?

Reply:
i don't think it can happen as currently coded, was my point, but it's possible there's a bug in the logic.

re: 4B, are you thinking we'd hit some memory limit or something? i haven't considered that, but it seems untenable with the current architecture.

Reply:
Eh, at this point in code reviews I'm just trying to get a sense for places where things will break.  Since this PR about a 65K limit with Phasers, it seems natural to ask where the next boundary is.

It seems like we have a Phaser of Phasers, right?  If that's correct, then if we spun up 65K+ Phasers, the outer Phaser would break, right?  And to get there, we'd 65K * 65 ~= 4 billion+ files in a directory?  

If my statement above is wrong, then that's fine.  My second question/point is basically, if someone gets 4 billion files in a directory, they may be beyond our assistance:) and a GeoMesa bug will be the least of their concerns.

Reply:
ah, gotcha. no, it shouldn't hit any phaser limits at the moment. iirc if it hits the limit it will use the last registration spot to create a child phaser, and then start using that (which will give another 65k registrations), and if it hits that limit it will keep creating more levels of child phasers, etc.